pkg/resource/deploy/plan_executor.go

```diff
@@ -0,0 +1,224 @@
+// as the glue that links the two subsystems together.
+type PlanExecutor struct {
+	plan    *Plan          // The plan that we are executing
+	opts    Options        // Options for the plan execution
+	src     SourceIterator // The iterator that generates SourceEvents
+	preview bool           // are we running a preview?
+
+	stepGen  *stepGenerator // step generator owned by this plan
+	stepExec *stepExecutor  // step executor owned by this plan
+
+	ctx            *cancel.Context // Ctrl-C cancellation context, given to us by our caller.
+	stepExecCancel *cancel.Source  // step executor cancellation context, shared ownership between step executor and us.
+
+	sawError  uint32 // have we seen an error?
+	sawCancel uint32 // have we seen a cancel?
+}
+
+// Execute executes a plan to completion, using the given cancellation context and running a preview
+// or update.
+func (pe *PlanExecutor) Execute() error {
+	// Before heading into the main event loop, we launch two goroutines. The first one links
+	// the parent cancellation context (which signals cancellation from the CLI level, i.e. Ctrl+C)
+	// to the cancellation context that the plan executor shares with its step executor. This ensures
+	// that top-level cancellations result in quick teardown of all worker threads and the plan executor
+	// itself.
+	go func() {
+		<-pe.ctx.Canceled()
+		logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): received cancel signal")
+		atomic.SwapUint32(&pe.sawCancel, 1)
+		pe.stepExecCancel.Cancel()
+	}()
+
+	// The second one polls for incoming source events and writes them to the `incomingEvents` channel,
+	// so the man loop can `select` on it.
+	type nextEvent struct {
+		Event SourceEvent
+		Error error
+	}
+
+	incomingEvents := make(chan nextEvent)
+	go func() {
+		for {
+			event, sourceErr := pe.src.Next()
+			incomingEvents <- nextEvent{event, sourceErr}
+		}
+	}()
+
+	// The main loop. We'll continuously select for incoming events and the cancellation signal. There are
+	// a three ways we can exit this loop:
+	//  1. The SourceIterator sends us a `nil` event. This means that we're done processing source events and
+	//     we should begin processing deletes.
+	//  2. The SourceIterator sends us an error. This means some error occurred in the source program and we
+	//     should bail.
+	//  3. The stepExecCancel cancel context gets canceled. This means some error occurred in the step executor
+	//     and we need to bail. This can also happen if the user hits Ctrl-C.
+outer:
+	for {
+		logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): waiting for incoming events")
+		select {
+		case event := <-incomingEvents:
+			logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): incoming event")
+			if event.Error != nil {
+				logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): saw incoming error: %v", event.Error)
+				pe.cancelDueToError()
```

Comment:
It wasn't evident to me, but I assume somehow the `event.Error` ends up propagated in a way the user will see? (I kind of expected `cancelDueToError` to take an `error` argument...)

Reply:
The plan executor and step executors are not responsible for reporting errors that they see unless the errors arose internally within those components. For example, the step executor will report any errors that come from pre or post-step events, as well as errors that can occur when handling `RegisterResourceOutputs`.

It's hard to reason about who is responsible for reporting what errors, but in general with this PR I tried to keep error reporting away from the core execution engine as much as possible. In this particular case, when we see an `Error` coming from a SourceIterator, the source iterator implementation itself has already reported the error and we don't need to do so again.